src/cloud-providers/byom/provider.go

```diff
@@ -0,0 +1,265 @@
+// (C) Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package byom
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"strings"
+	"time"
+
+	provider "github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util/cloudinit"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var logger = log.New(log.Writer(), "[adaptor/cloud/byom] ", log.LstdFlags|log.Lmsgprefix)
+
+const (
+	sshPort      = "22"
+	userDataFile = "/media/cidata/user-data" // User-data file
+	rebootFile   = "/media/cidata/reboot"    // Reboot trigger file
+)
+
+// byomProvider implements the Provider interface for BYOM
+type byomProvider struct {
+	serviceConfig *Config
+	globalPoolMgr GlobalVMPoolManager
+	sshConfig     *ssh.ClientConfig // Pre-computed SSH client configuration
+}
+
+// NewProvider creates a new BYOM provider instance
+func NewProvider(config *Config) (provider.Provider, error) {
+	logger.Printf("BYOM config: %+v", config.Redact())
+
+	// Initialize SSH configuration and keys
+	sshConfig := &util.SSHConfig{
+		PublicKey:           config.SSHPubKey,
+		PrivateKey:          config.SSHPrivKey,
+		PublicKeyPath:       config.SSHPubKeyPath,
+		PrivateKeyPath:      config.SSHPrivKeyPath,
+		Username:            config.SSHUserName,
+		Timeout:             time.Duration(config.SSHTimeout) * time.Second,
+		HostKeyAllowlistDir: config.SSHHostKeyAllowlistDir,
+		EnableSFTP:          true, // Always enabled for BYOM
+	}
+
+	// Create SSH client configuration (also initializes keys if needed)
+	sshClientConf, err := util.CreateSSHClient(sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create SSH client configuration: %w", err)
+	}
+
+	// Update config with initialized keys
+	config.SSHPubKey = sshConfig.PublicKey
```

Comment:
As we are adding the generated public key to `authorized_keys` file in the pod VM image to be used for verification, any particular reason for it to the provider as well?

Reply:
Both are there for consistency.